Port the full MISMIP+ test group from the old vesrion of compass (#737)
MPAS-Dev/compass: docs/developers_guide/landice/test_groups/mismipplus.rst

.. _dev_landice_mismipplus_framework:

Framework
---------

The shared config options for the ``mismipplus`` test group
are described in :ref:`landice_mismipplus` in the User's Guide.

Test cases within the ``mismipplus`` group should only be run with the 
FO velocity solver, or with a data velocity field. Running with the FO solver 
requires building MALI with the Albany library.

The test group has a shared ``namelist.landice`` file with
a few common namelist options related to model numerics,
and a shared ``streams.landice`` file that defines ``input``, ``restart``, and
``output``, and ``globalStatsOutPut`` streams. There also is an 
``albany_input.yaml`` file that is needed for running the FO solver.

setup_mesh
~~~~~~~~~~
The class :py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh` 
defines a step for creating the mesh/initial conditions for a MISMIP+ spinup 
simulation. This step is only used by the ``spin_up`` test case. An odd number 
of rows are culled along the North/South boundaries such that the ragged 
layout of mesh rows is symetric in the y direction. 

The ``resolution`` parameter from the ``[mesh]`` section of the configuration

> file is the nominal resolution, where the actual resolution is itteratively 

Reviewer: ```suggestion
file is the nominal resolution, where the actual resolution is iteratively 
```

 
run_model
~~~~~~~~~

The class :py:class:`compass.landice.tests.mismipplus.run_model.RunModel`
defines a step for setting up the input files (i.e. streams, namelist, 
``albany_input.yaml``) needed to run MALI. The initial codnition can either 
be downloaded from the shared data server (for the 
:ref:`landice_mismipplus_smoke_test` testcase) or generated by the 
:py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh` step 
(for the :ref:`landice_mismipplus_smoke_test` testcase).

.. _dev_landice_mismipplus_tasks:

tasks
~~~~~

The function :py:func:`compass.landice.tests.mismipplus.tasks.get_ntasks_from_cell_count()`
can be used to compute a good number of MPI tasks (both the target and the
minimum) for MALI to use based on the ``goal_cells_per_core`` and
``max_cells_per_core`` config options as well as the number of cells in a mesh.
The idea is that we want to run MALI with about 300 cells per core
(the default value of ``goal_cells_per_core``) but that we would be okay
with as many as 5000 cells per core (the default ``max_cells_per_core``).

A complication of using this function is that the number of cells in a mesh

> is not know unitl the :py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh`

Reviewer: ```suggestion
is not known until the :py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh`
```

step has been run (i.e. at ``compass run``) for a given ``resolution``, but we 
need to know how many cores and nodes we will use at setup 

> (i.e. at ``compass setup ...``) since thats when the job script is created. 

Reviewer: ```suggestion
(i.e. at ``compass setup ...``) since that is when the job script is created. 
```
